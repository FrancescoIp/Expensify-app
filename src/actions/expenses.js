import uuid from 'uuid';
import database from '../firebase/firebase';
import expenses from '../reducers/expenses';
// Prima di firebase-----
//componets calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//dopo firebase----------
//component call action generator
//action generator returns function (redux need a module as support middleware to dispatch functions)
//component dispatches function
//function runs( has ability to dispatch other action or what he wants )


// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt }
    // it seems that firebase can't handle this asynchronous code, so changing to sync code
    const ref = database.ref(`users/${uid}/expenses`).push(expense);
    dispatch(addExpense({
      id: ref.key,
      ...expense
    }))

    return ref
  };
};

// REMOVE_EXPENSE.
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    });
  };
};

// EDIT_EXPENSE. nessun valore di default perchÃ¨ non serve
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});


export const startEditExpense = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    });
  };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})


export const startSetExpenses = () => {
  return (dispatch, getState) => {
    //1. fetch all expenses of the logged user
    const uid = getState().auth.uid
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      //2. Parse expenses into an array
      const expenses = [];

      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      //3. dispatch setExpenses with that array
      dispatch(setExpenses(expenses));
    });
  };
};


