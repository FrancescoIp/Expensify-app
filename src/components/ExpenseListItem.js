import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, id}) => (
  <div>
    <li>
      <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
      <p>Amount: {amount/100}</p>
      <p>Created At: {createdAt}</p>
    </li>
  </div>
)


export default ExpenseListItem