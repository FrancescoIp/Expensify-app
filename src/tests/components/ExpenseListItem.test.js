import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixture/expenses';

test('should render an expense with givend data', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[1]}/>);
  expect(wrapper).toMatchSnapshot();
});