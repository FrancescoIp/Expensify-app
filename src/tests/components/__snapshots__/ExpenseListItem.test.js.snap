// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expense with givend data 1`] = `
<div>
  <li>
    <Link
      replace={false}
      to="/edit/2"
    >
      <h3>
        Rent
      </h3>
    </Link>
    <p>
      Amount: 
      $1,095.00
    </p>
    <p>
      Created At: 
      December 28th, 1969
    </p>
  </li>
</div>
`;
